descqa/AngularCorrelation.py

```diff
@@ -0,0 +1,180 @@
+        data = {}
+        if not gc.has_quantities(quantities):
+            return TestResult(skipped=True, summary='Missing requested quantities')
+
+        data = gc.get_quantities(quantities, filters=filters)
+        #make sure data entries are all finite
+        data = GCRQuery(*((np.isfinite, col) for col in data)).filter(data)
+
+        return data
+
+
+    def get_validation_data(self, observation, maglimit):
+        '''
+        Get the observations
+        '''
+        data_args = self.possible_observations[observation]
+        data_path = os.path.join(self.data_dir, data_args['filename_template'].format(maglimit))
+
+        if not os.path.exists(data_path):
+            raise ValueError("{}-maglimit data file {} not found".format(maglimit, data_path))
+
+        if not os.path.getsize(data_path):
+            raise ValueError("{}-maglimit data file {} is empty".format(maglimit, data_path))
+
+        data = np.loadtxt(data_path, unpack=True, usecols=data_args['usecols'], skiprows=data_args['skiprows'])
+
+        validation_data = dict(zip(data_args['colnames'], data))
+        validation_data['label'] = data_args['label']
+
+        return validation_data
+
+
+    def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
+        '''
+        Loop over magnitude cuts and make plots
+        '''
+        results = {}
+        filenames = ['r17_18.dat', 'r18_19.dat', 'r19_20.dat', 'r20_21.dat', 'r17_21.dat']
+        labels = [r'$r=17-18$', r'$r=18-19$', r'$r=19-20$', r'$r=20-21$', r'$r=17-21$']
+        colors = plt.cm.jet(np.linspace(0, 1, len(filenames)))
+        #Maximum magnitude, minimum luminosity
+        mag_r_maxs = [18, 19, 20, 21, 21]
+        #Minimum magnitude, maximum luminosity
+        mag_r_mins = [17, 18, 19, 20, 17]
+        min_sep = 1e-2 #Degree
+        max_sep = 5 #Degree
+        bin_size = 0.2
+        for rmax, rmin, fname, label, color in zip(mag_r_maxs, mag_r_mins, filenames, labels, colors):
+            #check for valid observations
+            maglimit = '{}_{}'.format(rmin, rmax) 
+            if not self.observation:
+                print('Warning: no data file supplied, no observation requested; only catalog data will be shown')
+            elif self.observation not in self.possible_observations:
+                raise ValueError('Observation {} not available'.format(self.observation))
+            else:
+                self.validation_data = self.get_validation_data(self.observation, maglimit)
+            #Constrain in redshift and magnitude. DESQA accepts it as a function of corresponding quantities (in this case redshift and r-band magnitude)
+            filters = [(lambda z: (z<self.redshift_max),'redshift'),(lambda mag: (mag>rmin)&(mag<rmax),self.mag_field)] 
+            #Generating catalog data based on the constrains and getting the columns ra and dec
+            catalog_data = self.get_catalog_data(catalog_instance, ['ra', 'dec'], filters=filters)
```

Comment:
Also, it feels like we can load the data (with only a filter on z) outside the loop, and then manually select the correct mag range within the loop.